--- qmail-smtpd.c	2018-04-02 19:31:50.973788625 -0600
+++ qmail-smtpd.c.new	2018-04-02 21:57:54.552345027 -0600
@@ -120,6 +120,7 @@
 }
 #endif
 void err_unimpl(arg) char *arg; { out("502 unimplemented (#5.5.1)\r\n"); }
+void err_unrecog() { out("500 unrecognized (#5.5.2)\r\n"); }
 void err_syntax() { out("555 syntax error (#5.5.4)\r\n"); }
 void err_wantmail() { out("503 MAIL first (#5.5.1)\r\n"); }
 void err_wantrcpt() { out("503 RCPT first (#5.5.1)\r\n"); }
@@ -1438,7 +1439,7 @@
 #endif
 , { "noop", err_noop, flush }
 , { "vrfy", err_vrfy, flush }
-, { 0, err_unimpl, flush }
+, { 0, err_unrecog, flush }
 } ;
 
 int main(argc,argv)
