--- qmail-remote.c	2018-04-02 22:08:17.572175085 -0600
+++ qmail-remote.c.new	2018-04-02 22:07:00.033859280 -0600
@@ -724,7 +724,8 @@
 #endif
  
   code = smtpcode();
-  if (code >= 500) quit("DConnected to "," but greeting failed");
+  if (code >= 500 && code < 600) quit("DConnected to "," but greeting failed");
+  if (code >= 400 && code < 500) return; /* try next MX, see RFC-2821 */
   if (code != 220) quit("ZConnected to "," but greeting failed");
  
 #ifdef EHLO
@@ -1007,6 +1008,7 @@
       partner_fqdn = ip.ix[i].fqdn;
 #endif
       smtp(); /* does not return */
+      smtp(); /* only returns when the next MX is to be tried */
     }
     tcpto_err(&ip.ix[i].ip,errno == error_timeout);
     close(smtpfd);
